xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java

```diff
@@ -82,19 +83,61 @@ protected void writeXml(XmlWriter writer) throws IOException {
          * and the application must take care of it.
          */
 
-        try (xmlInputStream; ) {
-            // fill the buffer once with the first 1024 chars and read as string
-            byte[] buffer = xmlInputStream.readNBytes(1024);
-            String firstChars = new String(buffer, StandardCharsets.UTF_8);
+        try (xmlInputStream;
+                // create a buffered stream to allow for rewind
+                BufferedInputStream bufferedXmlInputStream =
+                        new BufferedInputStream(xmlInputStream)) {
 
+            final int n = 1024;
+            // mark the start of the stream, leave 1 byte headroom for reading ahead
+            bufferedXmlInputStream.mark(n + 1);
+            // read the first bytes
+            byte[] bytes = bufferedXmlInputStream.readNBytes(n);
+
+            // broken UTF-8 multibyte char only possible when >= 1024 chars available
+            if (bytes.length == n) {
+                // analyze how far we can read without risk of broken char
+                final int maxN = maxBytesWithCompleteUTF8Chars(bytes);
+                // if we detected a (potentially broken) mbchar at the end, re-read.
+                if (maxN < n) {
+                    // reset the stream to begin
+                    bufferedXmlInputStream.reset();
+                    // replace bytes by reading again, but only so far as maxN
+                    bytes = bufferedXmlInputStream.readNBytes(maxN);
+                }
+            }
+
+            String firstChars = new String(bytes, StandardCharsets.UTF_8);
             // match the start with the compiled regex and replace with nothing when matching.
             firstChars = xmlDeclaration.reset(firstChars).replaceFirst("");
 
             // write the chars to the output stream
             writer.getOutputStream().write(firstChars.getBytes(StandardCharsets.UTF_8));
 
             // now send the rest of the stream
-            xmlInputStream.transferTo(writer.getOutputStream());
+            bufferedXmlInputStream.transferTo(writer.getOutputStream());
         }
     }
+
+    public static int maxBytesWithCompleteUTF8Chars(final byte[] buffer) {
+        final int n = buffer.length;
+
+        if (Byte.toUnsignedInt(buffer[n - 1])
+                > 127) { // 0b0xxxxxxx are all 1-byte chars, so only seek for 0b1xxxxxxx
+            for (int i = n - 1;
+                    i > n - 5;
+                    i--) { // go max four bytes back (max UTF-8 mbchar length)
+                if (Byte.toUnsignedInt(buffer[i])
+                        > 191) { // 0b110xxxxx / 0b1110xxxx / 0b11110xxx are the UTF-8 multibyte
```

Comment:
Just out of curiosity, was there anything objectively wrong with testing by converting the buffer to string, then checking if `firstChars.charAt(firstChars.length()-1) == '\uFFFD'`? - were there any cases of incomplete utf8 sequences that it wasn't going to cover?

Reply:
There definitely is some bikeshedding involved here. 

I didn't do a real performance test, but I had the nagging feeling that converting to string multiple times instead of looking at the encoded char might be slower.

Other than that, I liked following the UTF-8 encoding schema better, not relying on codepoint vs char etc.